### DISCUSSION TITLE
Nested Categories

### DISCUSSION THREAD

**Discussion by olivermrbl** [medusajs/medusa#2844] [Feature Requests]
### What
Introduce a concept Category in the products domain and allow merchants to create nested relationships

### Motivation
In our continuous efforts to extend our Products API, we will introduce categories. This will allow merchants to expose more granular customer-facing search capabilities. 

### Technical design
[RFC](https://github.com/medusajs/medusa/discussions/2873)

**Comment by iamanle** [comment]
hi @olivermrbl do you know when will this available?

**Comment by olivermrbl** [comment]
@anle1991  Work has significantly slowed down these days due to the Holidays, but we will pick up the work in the first week of 2023. I expect we will be running testing end of Jan or early Feb :)

**Comment by flashjames** [comment]
Answer to: @anle1991 "hi @olivermrbl do you know when will this available?" 

I need this too, and the base for this feature seem to have been merged to master:
https://github.com/medusajs/medusa/commits?author=riqwan&since=2023-02-01&until=2023-02-27

I don't know how complete it is yet, but maybe @riqwan or @olivermrbl can comment.

Note: Couldn't find any code in medusa-admin repo, so seems it's only backend for now, but that's good enough for me.

Edit: Found pull request for admin part: https://github.com/medusajs/admin/pull/890 status, not merged yet.

**Reply by olivermrbl** [reply]
Glad to hear there's demand for the feature @flashjames. All planned work for the core landed in `develop` this week, and the first iteration of the admin work is expected to be done next week. This is when we'll begin thorough testing of the feature.

If you want to use it now, I recommend installing the `@next` version of the core (and bump all related Medusa packages), as it contains a few PRs not part of `master` yet. Just note the `@next` version is the pre-release for v1.8.0, which contains breaking migrations e.g. to Typeorm 0.3.11, as well as many large features that are still very experimental. 

Additionally, you will have to enable the feature flag to use the store + admin API of Product Categories. Tweak your `medusa-config.js` to contain the following in its project configuration:
```
module.exports = {
  projectConfig: {
    ...
  },
  featureFlags: {
    product_categories: true,
  },
  plugins,
};
```

Please let me know if I can be of any assistance. And if you have feedback for the feature, we are all ears!

**Reply by flashjames** [reply]
I worked on some other stuff, but now I'm at this. First feedback is this @olivermrbl @riqwan 

### Add search for either handle or title
Right now there's only possible to search with q parameter (GET /admin/product-categories/?q=perfume), which searches both handle and name, should be able to search for either of them also. Maybe there's a way to make custom search queries on all api endpoints (i've seen some clues about this in the code, but cant find any documentation on this)

### Specify depth of a categories you want to search. 

For example, say we have this category subtree:
perfume/
perfume/children-perfume/

 ?depth=0&q=perfume , would return [perfume]
 ?depth=0,1&q=perfume , would return [perfume, children-perfume]
 ?depth=1&q=perfume , would return [children-perfume]

### Category metadata
It would be useful to be able to add metadata to a category, for example the description for the category when landing on it. Just like we have a metadata field on product

### A handle should be able to be used multiple times under different sub categories (i.e. not be unique
This is a bit more work, but just throwing it out there.

For example for this scenario:
perfume/children-perfume/armani
hair-care/shampoo/armani

This is not possible right now, since the last handle: "armani" is used in two child categories and a handle is unique globally.

**Reply by riqwan** [reply]
Thanks for the feedback @flashjames! 

**Add search for either handle or title**

I'll take a stab at this one soon to see if I can come up with something. 

**Specify depth of a categories you want to search.**

I had this in my plans, but had to abandon it because of how messy it gets with typeorm's tree repositories for list endpoints. At some point we do have to include a custom tree repository in the core that deals with how we can use tree queries properly for list endpoints. With the current approach, this could get quite messy. Happy to hear some ideas if you've got any. 

Can you share some use-cases on why you'd want to search for a category under a certain depth? I haven't seen many examples of categories being text-searched on an ecommerce store. 

**Category metadata**

I think this is a good call. I'll bring this back to the team.

**Handle case:**

You can handle this scenario right now by setting `perfume/children-perfume/armani` as a string to the categories handle. Any reason why this wouldn't work?

**Reply by flashjames** [reply]
> Thanks for the feedback @flashjames!
> 
> **Add search for either handle or title**
> 
> I'll take a stab at this one soon to see if I can come up with something.
> 
> **Specify depth of a categories you want to search.**
> 
> I had this in my plans, but had to abandon it because of how messy it gets with typeorm's tree repositories for list endpoints. At some point we do have to include a custom tree repository in the core that deals with how we can use tree queries properly for list endpoints. With the current approach, this could get quite messy. Happy to hear some ideas if you've got any.
> 
> Can you share some use-cases on why you'd want to search for a category under a certain depth? I haven't seen many examples of categories being text-searched on an ecommerce store.
> 
> **Category metadata**
> 
> I think this is a good call. I'll bring this back to the team.
> 
> **Handle case:**
> 
> You can handle this scenario right now by setting `perfume/children-perfume/armani` as a string to the categories handle. Any reason why this wouldn't work?

"Add search for either handle or title"
I moved around this problem, by just grabbing the whole categories GET /product-categories/
Since I needed to sort the categories by handle and by name for my mega menu, so it made sense right there to just download whole product category data, even if it's not optimal in loading time.

"Can you share some use-cases on why you'd want to search for a category under a certain depth? I haven't seen many examples of categories being text-searched on an ecommerce store."
I dont remember why I needed it.

"Handle case:"
Yeah that could work, but then I would be good if categories could have multiple handles.

## Storefront product-categories and filtering - Suggestion

By the way, I have soon finished the code for the storefront for displaying categories and filtering them.
I don't have the time to make it generic and make a pull request (atleast right now) and I wont make the code public.

But if you or another teammate want re-use my code or at least have something to have as a starting point, I could invite you or your teammate to my repository.

What I can see, you seem to prioritize the backend and work on the storefront later when you have improved the backend which makes sense. To reuse parts of my code would make it fast to get this feature into the storefront which kind of is a must have for many stores.


@riqwan @olivermrbl I can definitely help making the storefront code more generic, but I don't have time right now to do it on my own.

The UI filtering code is based on: https://tailwindui.com/components/ecommerce/components/category-filters

![image](https://user-images.githubusercontent.com/359702/226548876-c633e1a6-f651-472d-bb06-5983ac10809b.png)

## Current limitations of the GET /store/products (list and filter)
A few fields I would like to add filters for is:
- Brand
- Price range
- And other differents "product tags"

### For brand
Right now there's no way to get product tags for a product category, for example I would need all brands for a product category and it would be helpful to get a list of number of products that has that tag under a product category.
But maybe this is something that shouldn't be part of core and that you should implement  your own logic for. But I think it should be possible to make it more generic, but this would probably mean you need to make custom fields on products and not save it as tags or as metadata. Anyways, just throwing out my initial thoughts.

I guess it's a limitation on:
- Custom fields on products
- Custom fields on tags (right now you can only have a value for a tag I think)
I think I've read somewhere on the roadmap that this is something your working on.

EDIT: Found the discussion https://github.com/medusajs/medusa/discussions/1947

[image](https://user-images.githubusercontent.com/359702/226549806-3f104f66-6ac6-47d2-a75e-fb9e762f2d8d.png)

### For price range
There's no way to limit products by price range what I can see.
